### PR TITLE
[SPARK-15915][SQL] Logical plans should use subqueries eliminated plan when override sameResult.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

This pr is a backport of #13638 for `branch-1.6`.
## How was this patch tested?

Added the same test as #13638 modified for `branch-1.6`.

### CODE REVIEW

**Comment by cloud-fan** [comment]
LGTM

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by ueshin** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala [hunk 1]
```diff
@@ -56,10 +57,12 @@ case class LocalRelation(output: Seq[Attribute], data: Seq[InternalRow] = Nil)
 
   override protected def stringArgs = Iterator(output)
 
-  override def sameResult(plan: LogicalPlan): Boolean = plan match {
-    case LocalRelation(otherOutput, otherData) =>
-      otherOutput.map(_.dataType) == output.map(_.dataType) && otherData == data
-    case _ => false
+  override def sameResult(plan: LogicalPlan): Boolean = {
+    EliminateSubQueries(plan) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala]: Maybe we should execute `EliminateSubQueries` when looking up cache entry? 1.6 is not like 2.0, we don't have the `canonized` method and can't memorize the result of `EliminateSubQueries`.

Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala]: I'm sorry but I couldn't find the reason.
Is the purpose to execute `EliminateSubQueries` to memorize the result of it?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala]: nvm, I was wrong

**Comment by cloud-fan** [comment]
thanks, merging to 1.6!

@ueshin can you close this PR? it won't be closed automatically.

**Comment by ueshin** [comment]
@cloud-fan Sure, thank you for merging this!

**Pull request closed without merging**